lib/QoreSerializable.cpp

```diff
@@ -0,0 +1,1711 @@
+
+    ConstHashIterator hi(h);
+    while (hi.next()) {
+        // write key string to stream
+        if (serializeStringToStream(writer, hi.getKey(), strlen(hi.getKey()), QCS_DEFAULT, xsink)) {
+            return -1;
+        }
+
+        if (serializeValueToStream(hi.get(), writer, xsink)) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+int QoreSerializable::serializeStringToStream(const QoreStringNode& str, StreamWriter& writer, ExceptionSink* xsink) {
+    return serializeStringToStream(writer, str.c_str(), str.size(), str.getEncoding(), xsink);
+}
+
+int QoreSerializable::serializeStringToStream(StreamWriter& writer, const char* str, size_t len, const QoreEncoding* enc, ExceptionSink* xsink) {
+    // write string encoding if not UTF-8
+    if (enc && enc != QCS_UTF8) {
+        // write data type code to stream
+        if (writer.writei1(QSSDT_STRING, xsink)) {
+            return -1;
+        }
+        // get encoding string
+        const char* enc_str = enc->getCode();
+        size_t enc_len = strlen(enc_str);
+
+        // write encoding string size to stream
+        if (QoreSerializable::serializeIntToStream(enc_len, writer, xsink)) {
+            return -1;
+        }
+
+        // write string
+        if (writer.write(enc_str, enc_len, xsink)) {
+            return -1;
+        }
+    }
+    else {
+        // write data type code to stream
+        if (writer.writei1(QSSDT_UTF8_STRING, xsink)) {
+            return -1;
+        }
+    }
+
+    // write string size
+    if (QoreSerializable::serializeIntToStream(len, writer, xsink)) {
+        return -1;
+    }
+
+    // write string
+    return writer.write(str, len, xsink);
+}
+
+int QoreSerializable::serializeIntToStream(int64 i, StreamWriter& writer, ExceptionSink* xsink) {
+    if ((i >= -128) & (i <= 127)) {
+        // write data type code to stream
+        if (!writer.writei1(QSSDT_INT1, xsink)) {
+            // write integer to stream
+            return writer.writei1((signed char)i, xsink);
+        }
+    }
+    else if ((i >= -32768) && (i <= 32767)) {
+        // write data type code to stream
+        if (!writer.writei1(QSSDT_INT2, xsink)) {
+            // write integer to stream
+            return writer.writei2((int16_t)i, xsink);
+        }
+    }
+    else if ((i >= -2147483648) && (i <= 2147483647)) {
+        // write data type code to stream
+        if (!writer.writei1(QSSDT_INT4, xsink)) {
+            // write integer to stream
+            return writer.writei4((int32_t)i, xsink);
+        }
+    }
+    else {
+        // write data type code to stream
+        if (!writer.writei1(QSSDT_INT8, xsink)) {
+            // write integer to stream
+            return writer.writei8((int64_t)i, xsink);
+        }
+    }
+
+    return -1;
+}
+
+int QoreSerializable::serializeBoolToStream(bool b, StreamWriter& writer, ExceptionSink* xsink) {
+    // write data type code to stream
+    if (!writer.writei1(QSSDT_BOOLEAN, xsink)) {
```

Comment:
1-byte per bool could be saved if there was simply constants `QSSDT_BOOLEAN_TRUE` and `QSSDT_BOOLEAN_FALSE`.

Reply:
Also the code overhead would be smaller.